### PR TITLE
KOGITO-6807: Implement undo, redo and isDirty on the Serverless Workflow Editor

### PR DESCRIPTION
**Pull request by paulovmr**

### CODE REVIEW

File: packages/serverless-workflow-editor/src/editor/ServerlessWorkflowEditor.tsx [hunk 1]
```diff
@@ -103,33 +114,46 @@ const RefForwardingServerlessWorkflowEditor: React.ForwardRefRenderFunction<
         },
       };
     },
-    [content]
+    [monacoEditorRef]
```
Comment (tiagobento) [inline packages/serverless-workflow-editor/src/editor/ServerlessWorkflowEditor.tsx]: Refs don't need to be dependencies, as React guarantees that they're reference is unchanged.

File: packages/serverless-workflow-editor/src/editor/ServerlessWorkflowEditor.tsx [hunk 2]
```diff
@@ -37,6 +39,13 @@ interface Props {
    */
   ready: () => void;
 
+  /**
+   * Delegation for KogitoEditorChannelApi.kogitoEditor_stateControlCommandUpdate(command) to signal to the Channel
+   * that the editor is performing an undo/redo operation. Increases the decoupling of the ServerlessWorkflowEditor
+   * from the Channel.
+   */
+  stateControlCommandUpdate: (command: StateControlCommand) => void;
```
Comment (tiagobento) [inline packages/serverless-workflow-editor/src/editor/ServerlessWorkflowEditor.tsx]:
```suggestion
  onStateControlCommandUpdate: (command: StateControlCommand) => void;
```

File: packages/serverless-workflow-editor/src/editor/ServerlessWorkflowEditor.tsx [hunk 3]
```diff
@@ -37,6 +39,13 @@ interface Props {
    */
   ready: () => void;
```
Comment (tiagobento) [inline packages/serverless-workflow-editor/src/editor/ServerlessWorkflowEditor.tsx]: on
```suggestion
  onReady: () => void;
```

Comment (tiagobento) [reply packages/serverless-workflow-editor/src/editor/ServerlessWorkflowEditor.tsx]: and `onNewEdit` too.

File: packages/serverless-workflow-editor/src/editor/ServerlessWorkflowEditor.tsx [hunk 4]
```diff
@@ -103,33 +114,46 @@ const RefForwardingServerlessWorkflowEditor: React.ForwardRefRenderFunction<
...
   );
 
-  useEffect(() => {
-    props.ready();
-  }, [originalContent, props]);
+  const contentChanged = useCallback(
+    (newContent: string, operation?: MonacoEditorOperation) => {
+      if (operation === MonacoEditorOperation.EDIT) {
+        props.newEdit(new KogitoEdit(newContent));
+        monacoEditorRef.current?.pushUndoStop();
```
Comment (tiagobento) [inline packages/serverless-workflow-editor/src/editor/ServerlessWorkflowEditor.tsx]: Do you need to call this here? I mean, this seems like a piece of logic that could be moved to inside the Monaco editor component at the same place you call `onContentChanged` with the operation equal to MonacoEditorOperation.EDIT. This way we avoid adding this `pushUndoStop` method to the ref interface and simplify things. WDYT?

File: packages/serverless-workflow-editor/src/editor/ServerlessWorkflowEditor.tsx [hunk 5]
```diff
@@ -103,33 +114,46 @@ const RefForwardingServerlessWorkflowEditor: React.ForwardRefRenderFunction<
...
       };
     },
-    [content]
+    [monacoEditorRef]
   );
 
-  useEffect(() => {
-    props.ready();
-  }, [originalContent, props]);
+  const contentChanged = useCallback(
```
Comment (tiagobento) [inline packages/serverless-workflow-editor/src/editor/ServerlessWorkflowEditor.tsx]:
```suggestion
  const onContentChanged = useCallback(
```

File: packages/serverless-workflow-editor/src/monaco/augmentation/MonacoEditorApi.ts [hunk 6]
```diff
@@ -23,28 +23,53 @@ export interface MonacoEditorApi {
...
+  constructor(
+    content: string,
+    onContentChange: (content: string, operation: MonacoEditorOperation) => void,
+    augmentation: MonacoAugmentation
+  ) {
     this.model = editor.createModel(augmentation.language.getDefaultContent(content), augmentation.language.languageId);
-    this.model.onDidChangeContent((event) => onContentChange(this.model.getValue()));
+    this.model.onDidChangeContent((event) => {
+      if (!event.isUndoing && !event.isRedoing) {
+        onContentChange(this.model.getValue(), MonacoEditorOperation.EDIT);
```
Comment (tiagobento) [inline packages/serverless-workflow-editor/src/monaco/augmentation/MonacoEditorApi.ts:55]: Here I think you could call the `pushUndoStop`. Would that work?

File: packages/serverless-workflow-editor/src/monaco/augmentation/MonacoEditorApi.ts [hunk 7]
```diff
@@ -23,28 +23,53 @@ export interface MonacoEditorApi {
...
+      }
+    });
+    this.onContentChange = onContentChange;
     this.augmentation = augmentation;
   }
 
   redo(): void {
+    this.editor?.focus();
     this.editor?.trigger("whatever...", "redo", null);
```
Comment (tiagobento) [inline packages/serverless-workflow-editor/src/monaco/augmentation/MonacoEditorApi.ts]:
```suggestion
    this.editor?.trigger("??????", "redo", null);
```

What's the first parameter here? I think we could use something better than `whatever` 😝

File: packages/serverless-workflow-editor/src/monaco/augmentation/MonacoEditorApi.ts [hunk 8]
```diff
@@ -23,28 +23,53 @@ export interface MonacoEditorApi {
...
   }
 
   undo(): void {
+    this.editor?.focus();
     this.editor?.trigger("whatever...", "undo", null);
```
Comment (tiagobento) [inline packages/serverless-workflow-editor/src/monaco/augmentation/MonacoEditorApi.ts]: Same here.

File: packages/serverless-workflow-editor/src/monaco/augmentation/MonacoEditorApi.ts [hunk 9]
```diff
@@ -57,7 +82,21 @@ export class DefaultMonacoEditor implements MonacoEditorApi {
...
+    });
+
+    this.editor.addCommand(KeyMod.CtrlCmd | KeyCode.KEY_Y, () => {
+      this.onContentChange(this.model.getValue(), MonacoEditorOperation.REDO);
+    });
+
+    this.editor.addCommand(KeyMod.CtrlCmd | KeyMod.Shift | KeyCode.KEY_Z, () => {
+      this.onContentChange(this.model.getValue(), MonacoEditorOperation.REDO);
+    });
+  }
```
Comment (tiagobento) [inline packages/serverless-workflow-editor/src/monaco/augmentation/MonacoEditorApi.ts]: This is a bit controversial. I think we should base the decision on the current OS. What's the default keybinding for `undo` on VS Code for Windows, macOS and Linux? We should follow the same approach, I believe. Even better if we could get the associated key bindings with the `undo`/`redo` commands in VS Code.

**Review by tiagobento — CHANGES_REQUESTED** [review]
Nice PR, @paulovmr!

Left some comments inline, please take a look at them.

**Review by tiagobento — APPROVED** [review]

**Review by pefernan — APPROVED** [review]
+1

**Pull request merged**